toolchain/check/import_cpp.cpp

```diff
@@ -70,13 +68,85 @@ auto ImportCppFiles(
         IntAsSelect, IntAsSelect, IntAsSelect, std::string);
     context.emitter().Emit(loc, CppInteropParseError, num_errors, num_warnings,
                            num_imports, diagnostics_str);
-  } else if (num_warnings > 0) {
+    return std::nullopt;
+  }
+  if (num_warnings > 0) {
     CARBON_DIAGNOSTIC(CppInteropParseWarning, Warning,
                       "{0} warning{0:s} in `Cpp` {1} import{1:s}:\n{2}",
                       IntAsSelect, IntAsSelect, std::string);
     context.emitter().Emit(loc, CppInteropParseWarning, num_warnings,
                            num_imports, diagnostics_str);
   }
+  return ast;
+}
+
+static auto AddNamespace(Context& context, IdentifierId cpp_package_id,
+                         SemIR::InstId first_import_decl_id) -> void {
```

Comment:
Had you tried something like using `import.cpp`'s `AddNamespace`, i.e.:

```
AddNamespace(context, namespace_type_id, name_id, SemIR::NameScopeId::Package, /*diagnose_duplicate_namespace=*/true, [&] { return first_import_decl_id; });
```

What was the difference between that and what you're trying to do here?

Reply:
I've changed it further, but I still believe you're right that there are a lot of similarities and that's why I have a TODO extract and deduplicate the common logic.
Once we agree on the functionality, I can do this TODO as a follow up PR or as part of this PR.

Reply:
Sorry for missing this.
I've taken `AddNamespace()`, exposed it in `import.h` as `AddImportNamespace()` and used it in `import_cpp.cpp`.
I can put it elsewhere so I didn't see a great candidate except creating a dedicated file. WDYT?

It seems to work fine though there are some impact on the behavior in api and impl use case, due to how duplicated namespace are handled (which might be a bug fix?).

Reply:
I think `import.*` is still the right home for the function; the import_cpp -> import dependency is good.

Yes, I think this is pushing in the right direction for merging imports. Note, you might want a TODO to handle merging `api` and `impl` ASTs... there are a few options there though.

Reply:
Why is this dependency good? Seems like the new utility function can be extracted and used by the implementations of the two APIs, so we won't need to have one API depend on the other.

I think merging AST (and a TODO) belongs to a PR that actually tries to save the ASTs for later usage.